yarn-project/ethereum/src/gas_utils.ts

```diff
@@ -0,0 +1,216 @@
+   */
+  public async getGasPrice(): Promise<bigint> {
+    const block = await this.publicClient.getBlock({ blockTag: 'latest' });
+    const baseFee = block.baseFeePerGas ?? 0n;
+    const priorityFee = this.gasConfig.priorityFeeGwei * WEI_CONST;
+
+    const baseWithBuffer = baseFee + (baseFee * this.gasConfig.bufferPercentage) / 100n;
+    const totalGasPrice = baseWithBuffer + priorityFee;
+
+    const maxGasPrice = this.gasConfig.maxGwei * WEI_CONST;
+    const minGasPrice = this.gasConfig.minGwei * WEI_CONST;
+
+    let finalGasPrice: bigint;
+    if (totalGasPrice < minGasPrice) {
+      finalGasPrice = minGasPrice;
+    } else if (totalGasPrice > maxGasPrice) {
+      finalGasPrice = maxGasPrice;
+    } else {
+      finalGasPrice = totalGasPrice;
+    }
+
+    this.logger?.debug(
+      `Gas price calculation: baseFee=${baseFee}, withBuffer=${baseWithBuffer}, priority=${priorityFee}, final=${finalGasPrice}`,
+    );
+
+    return finalGasPrice;
+  }
+
+  /**
+   * Estimates gas with retries and adds buffer
+   */
+  public estimateGas<T extends (...args: any[]) => Promise<bigint>>(estimator: T): Promise<bigint> {
+    return retry(
+      async () => {
+        const gas = await estimator();
+        return gas + (gas * this.gasConfig.bufferPercentage) / 100n;
+      },
+      'gas estimation',
+      makeBackoff([1, 2, 3]),
+      this.logger,
+      false,
+    );
+  }
+
+  /**
+   * Monitors a transaction and attempts to speed it up if it gets stuck
+   * @param txHash - The hash of the transaction to monitor
+   * @param walletClient - The wallet client for sending replacement tx
+   * @param originalTx - The original transaction data for replacement
+   * @param config - Monitor configuration
+   */
+  public async monitorTransaction(
+    txHash: Hex,
+    walletClient: WalletClient,
+    originalTx: {
+      to: Address;
+      data: Hex;
+      nonce: number;
+      gasLimit: bigint;
+      maxFeePerGas: bigint;
+    },
+    txMonitorConfig?: TransactionMonitorConfig,
+  ): Promise<TransactionReceipt> {
+    const config = { ...this.monitorConfig, ...txMonitorConfig };
+    let attempts = 0;
+    let lastSeen = Date.now();
+    let currentTxHash = txHash;
+
+    // Flag to use for first run to avoid logging errors.
+    let firstRun = true;
+
+    while (true) {
+      try {
+        // Check transaction status
+        const receipt = await this.publicClient.getTransactionReceipt({ hash: currentTxHash });
+        if (receipt) {
+          this.logger?.info(`Transaction ${currentTxHash} confirmed`);
+          return receipt;
+        }
+
+        if (firstRun) {
+          firstRun = false;
+          // Wait a second for tx to be broadcast
+          await sleep(1000);
+        }
+
+        // Check if transaction is pending
+        const tx = await this.publicClient.getTransaction({ hash: currentTxHash });
+        this.logger?.info(`Transaction ${currentTxHash} pending`);
```

Comment:
You can do the same trick as with the receipt here. Instead of doing `getTx` on every hash you've sent, you can query `getTxCount` using the `pending` tag, which should consider the txs that are sitting on the mempool for the given sender.